crates/circuit/src/symbol_expr.rs

```diff
@@ -0,0 +1,3252 @@
+                match t.opt_complex() {
+                    Some(v) => v,
+                    None => t,
+                }
+            }
+        }
+    }
+    pub fn acos(&self) -> Value {
+        match self {
+            Value::Real(e) => Value::Real(e.acos()),
+            Value::Int(e) => Value::Real((*e as f64).acos()),
+            Value::Complex(e) => {
+                let t = Value::Complex(e.acos());
+                match t.opt_complex() {
+                    Some(v) => v,
+                    None => t,
+                }
+            }
+        }
+    }
+    pub fn tan(&self) -> Value {
+        match self {
+            Value::Real(e) => Value::Real(e.tan()),
+            Value::Int(e) => Value::Real((*e as f64).tan()),
+            Value::Complex(e) => {
+                let t = Value::Complex(e.tan());
+                match t.opt_complex() {
+                    Some(v) => v,
+                    None => t,
+                }
+            }
+        }
+    }
+    pub fn atan(&self) -> Value {
+        match self {
+            Value::Real(e) => Value::Real(e.atan()),
+            Value::Int(e) => Value::Real((*e as f64).atan()),
+            Value::Complex(e) => {
+                let t = Value::Complex(e.atan());
+                match t.opt_complex() {
+                    Some(v) => v,
+                    None => t,
+                }
+            }
+        }
+    }
+    pub fn exp(&self) -> Value {
+        match self {
+            Value::Real(e) => Value::Real(e.exp()),
+            Value::Int(e) => Value::Real((*e as f64).exp()),
+            Value::Complex(e) => {
+                let t = Value::Complex(e.exp());
+                match t.opt_complex() {
+                    Some(v) => v,
+                    None => t,
+                }
+            }
+        }
+    }
+    pub fn log(&self) -> Value {
+        match self {
+            Value::Real(e) => Value::Real(e.ln()),
+            Value::Int(e) => Value::Real((*e as f64).ln()),
+            Value::Complex(e) => {
+                let t = Value::Complex(e.ln());
+                match t.opt_complex() {
+                    Some(v) => v,
+                    None => t,
+                }
+            }
+        }
+    }
+    pub fn sqrt(&self) -> Value {
+        match self {
+            Value::Real(e) => {
+                if *e < 0.0 {
+                    Value::Complex(Complex64::from(e).powf(0.5))
+                } else {
+                    Value::Real(e.sqrt())
+                }
+            }
+            Value::Int(e) => {
+                if *e < 0 {
+                    Value::Complex(Complex64::from(*e as f64).powf(0.5))
+                } else {
+                    let t = (*e as f64).sqrt();
+                    let d = t.floor() - t;
+                    if (-f64::EPSILON..f64::EPSILON).contains(&d) {
+                        Value::Int(t as i64)
+                    } else {
+                        Value::Real(t)
```

Comment:
This isn't really correct floating-point handling.  The floating-point epsilon is very much a relative tool: it's very specifically just the smallest floating-point number such that `1. + e != 1.`.  It's meaningless if `t` is not close to 1, and even if it _is_, the number of floating-point values in `[t, t + e)` and `(t - e, t]` is not the same if `t` is a power of two, and I'm not certain what the test is asserting here.  If it's meant to be relative, then presumably what it actually wants is something related to the ULP difference between the two numbers, though against ofc there's a bunch of subnormal / infinite / NaN handling that may need to be done.

(Also, it's using a half-open range for the comparison, so it's asymmetric in another way too.)